### PR TITLE
Processors math

### PR DESCRIPTION
**Pull request by vi4eslav**
For partial preprocessing of a large volume of metrics, I added mathematical functions.

### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

**Comment by vi4eslav** [comment]
is there any news?

File: plugins/processors/math/README.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# Math Processor Plugin
+
+The math processor plugin applies the function to all or selected fields.
+
+### Configuration:
+
+```toml
+## Example config that processe all fields of the metric.
+# [[processor.math]]
```
Comment (ssoroka) [inline plugins/processors/math/README.md:9]: should be `[[processors.math]]`, here and other places it's mentioned.

File: plugins/processors/math/math.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+	return sampleConfig
+}
+
+func (p *Math) Description() string {
+	return "Math metrics that pass through this filter."
+}
+
+func (p *Math) Apply(in ...telegraf.Metric) []telegraf.Metric {
+
+	if _, ok := mathFunctions[p.Func]; ok {
```
Comment (ssoroka) [inline plugins/processors/math/math.go:100]: invert this to do an error check and flatten the rest of the function. Save the mathFunc reference.

File: plugins/processors/math/math.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+		for _, metric := range in {
+			if metric.Name() == p.Metric {
```
Comment (ssoroka) [inline plugins/processors/math/math.go:102]: FYI you can invert the check to bring the nesting and complexity down in cases like this:
```
if metric.Name() != p.Metric {
  continue
}
// now you don't need the nesting.
```
However, you probably want to drop this completely in favour of the built-in `namepass` feature, as this is more flexible, and it's handled for you already. see [metric filtering](https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md#metric-filtering)

File: plugins/processors/math/math.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/processors"
+)
+
+// Math function congfig
+type Math struct {
+	Func   string   `toml:"function"`
+	Metric string   `toml:"measurement_name"`
+	Fields []string `toml:"fields"`
```
Comment (ssoroka) [inline plugins/processors/math/math.go:16]: might want to also take a float64 argument variable, which would let you support the entire _math_ package, as well as addition, subtraction, multiplication, and division

File: plugins/processors/math/math.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+func (p *Math) Apply(in ...telegraf.Metric) []telegraf.Metric {
+
+	if _, ok := mathFunctions[p.Func]; ok {
+		for _, metric := range in {
+			if metric.Name() == p.Metric {
+				mathFunc := mathFunctions[p.Func]
+				if len(p.Fields) > 0 {
+					for _, name := range p.Fields {
+						if srcField, ok := metric.Fields()[name]; ok {
+							field, ok := srcField.(float64)
```
Comment (ssoroka) [inline plugins/processors/math/math.go:107]: can this be other types?

File: plugins/processors/math/math.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+func (p *Math) Apply(in ...telegraf.Metric) []telegraf.Metric {
+
+	if _, ok := mathFunctions[p.Func]; ok {
+		for _, metric := range in {
+			if metric.Name() == p.Metric {
+				mathFunc := mathFunctions[p.Func]
+				if len(p.Fields) > 0 {
+					for _, name := range p.Fields {
+						if srcField, ok := metric.Fields()[name]; ok {
```
Comment (ssoroka) [inline plugins/processors/math/math.go:106]: add a function that returns true when the metric has the right name and fields, then simplify this code to remove all the "is this the right metric" checks.

File: plugins/processors/math/math.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+							field, ok := srcField.(float64)
+							if ok {
+								metric.AddField(
+									fmt.Sprintf("%s_%s", name, p.Func),
+									mathFunc(field))
+							}
+						} else {
+							log.Printf(
+								"E! Metric %s doesn't have field %s",
```
Comment (ssoroka) [inline plugins/processors/math/math.go:115]: remove this. not useful. Most metrics won't have the right fields, and we just ignore it and continue on.

File: plugins/processors/math/math.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+								p.Func,
+								metric.Name())
+						}
+					}
+				} else {
+					for name, srcField := range metric.Fields() {
```
Comment (ssoroka) [inline plugins/processors/math/math.go:121]: remove this. if no fields are specified the function should do nothing.

File: plugins/processors/math/math.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+							metric.AddField(
+								fmt.Sprintf("%s_%s", name, p.Func),
+								mathFunc(field))
+						}
+					}
+				}
+			}
+		}
+	} else {
+		log.Printf(
```
Comment (ssoroka) [inline plugins/processors/math/math.go:133]: this should be at the top and return. Avoid `else`, especially when buried many lines down.

**Comment by ssoroka** [comment]
Discussed this with the team and we're going to close this in favor of the [Starlark processor plugin](https://github.com/influxdata/telegraf/tree/master/plugins/processors/starlark), which is already merged, and can handle more complex calculations. 
https://github.com/influxdata/telegraf/tree/master/plugins/processors/starlark

**Pull request closed without merging**